### ISSUE TITLE
[TF 2.1rc0] The `{predict,train,test}_on_batch` trace functions with fixed batch size

### ISSUE THREAD

**Issue by foxik** [tensorflow/tensorflow#34907]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Debian Stretch
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.1.0rc0
- Python version: 3.5.3

**Describe the current behavior**

When using the updated `predict_on_batch` method wrapped now in `tf.function`, the traced calls have a fixed batch size. Therefore, if batch size changes regularly, the function is retraced all the time. For example, the following

```python
import numpy as np
import tensorflow as tf

model = tf.keras.Sequential([
    tf.keras.layers.Dense(1, input_shape=[1])
])
model.compile(optimizer=tf.optimizers.Adam(),
              loss=tf.losses.MeanSquaredError())

for i in range(1, 300):
    model.predict_on_batch(np.ones([i, 1]))
```

gives

```
WARNING:tensorflow:5 out of the last 5 calls to <tensorflow.python.keras.engine.sequential.Sequential object at 0x7f2185a37630> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
WARNING:tensorflow:6 out of the last 6 calls to <tensorflow.python.keras.engine.sequential.Sequential object at 0x7f2185a37630> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.
...
```

The same issue happend with `train_on_batch` and `test_on_batch`.

**Describe the expected behavior**

The `{predict,train,test}_on_batch` methods are traced with undefined batch size.

Alternatively, the behaviour could be configurable. Each of `{predict,train,test}_on_batch` could take `undefined_batch_size=True/False` argument; personally I vote for a default value of `undefined_batch_size=True`.

**And I just realized:** if `*_on_batch` methods are used with sequences (for example sentences in NLP), then the retracing would happen on every call. So instead, it would be better to allow passing `experimental_relax_shapes` to the `*_on_batch` or even to `.compile` call.

**Code to reproduce the issue**
Available above.

**Other info / logs**

**Comment by oanush** [comment]
Issue replicating for the given code for TF version 2.1rc0, kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/1e2b044a36f464442c0c5d41437de4fc/34907.ipynb) of colab.Thanks!

**Comment by robieta** [comment]
That's not good. 

Thanks for providing such a simple and clean repro of the issue. You are correct; passing `experimental_relax_shapes` when we make the function (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training_v2_utils.py#L122) resolves the issue. Would you like to make the fix PR? (Otherwise I'm happy to do it.)

**Comment by foxik** [comment]
I have created the PR -- however it is just a "raw" change. Maybe some tests are also required? But I do not know to do them.

Also, the PR is against the master, but maybe it should be backported for TF2.1? Otherwise the final TF2.1 will have suboptimal _on_batch methods.

**Comment by robieta** [comment]
PR, for reference: https://github.com/tensorflow/tensorflow/pull/35019

@goldiegadde @martinwicke For getting it into 2.1
@yhliang2018 @bitfort We should be able to test this in microbenchmarks. (`experimental_relax_shapes` doesn't change semantics, only performance.)

**Comment by goldiegadde** [comment]
@foxik once the PR is merged into master, please request a CP to r2.1 branch.

**Comment by foxik** [comment]
@goldiegadde The PR has been merged into master as  f835a4a, so I am asking for a CP to r2.1. Is it enough to ask here or should I do something else?

**Comment by robieta** [comment]
I can handle the cherrypick. (I have another cherrypick request anyway, so adding one more is not any more work.) Thanks for the report and fix; this will greatly improve 2.1.

**Comment by LaurensHagendoorn** [comment]
**System information**

- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: No
- TensorFlow installed from (source or binary): Binary
- TensorFlow version (use command below): 2.2.0rc2
- Python version: 3.7.4

With the code above, I get the retracing warnings again in TF2.2.0rc2:
WARNING:tensorflow:11 out of the last 11 calls to <function Model.make_predict_function.<locals>.predict_function at 0x7fbc0061f5f0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings is likely due to passing python objects instead of tensors. Also, tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. Please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for more details.

**Comment by foxik** [comment]
@robieta Unfortunately, it again does not work with TF 2.2.0rc3

**Comment by robieta** [comment]
@foxik I'm no longer part of TensorFlow, but I'm sure @omalleyt12 can help you out. (This seems to have been lost when https://github.com/tensorflow/tensorflow/commit/10666c59dd4858645d1b03ce01f4450da80710ec dropped `training_v2_utils.py`, from the cursory skim that I did.)

**Comment by hiraksarkar** [comment]
Any update on the issue?
